# Delhivery - Feature engg

In [2]:
import  pandas as pd
import numpy as np



In [3]:
df= pd.read_csv('https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/001/551/original/delhivery_data.csv?')

In [4]:
df.head(10)

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   
5  training  2018-09-20 02:35:36.476840   
6  training  2018-09-20 02:35:36.476840   
7  training  2018-09-20 02:35:36.476840   
8  training  2018-09-20 02:35:36.476840   
9  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
5  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
6  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
7  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
8  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
9  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                    source_name  \
0  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
5  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
6  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
7  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
8  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
9  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
5       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
6       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
7       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
8       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
9       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   

                od_start_time  ...            cutoff_timestamp  \
0  2018-09-20 03:21:32.418600  ...         2018-09-20 04:27:55   
1  2018-09-20 03:21:32.418600  ...         2018-09-20 04:17:55   
2  2018-09-20 03:21:32.418600  ...  2018-09-20 04:01:19.505586   
3  2018-09-20 03:21:32.418600  ...         2018-09-20 03:39:57   
4  2018-09-20 03:21:32.418600  ...         2018-09-20 03:33:55   
5  2018-09-20 04:47:45.236797  ...         2018-09-20 06:15:58   
6  2018-09-20 04:47:45.236797  ...         2018-09-20 05:47:29   
7  2018-09-20 04:47:45.236797  ...         2018-09-20 05:25:58   
8  2018-09-20 04:47:45.236797  ...         2018-09-20 05:15:56   
9  2018-09-20 04:47:45.236797  ...         2018-09-20 04:49:20   

   actual_distance_to_destination  actual_time  osrm_time osrm_distance  \
0                       10.435660         14.0       11.0       11.9653   
1                       18.936842         24.0       20.0       21.7243   
2                       27.637279         40.0       28.0       32.5395   
3                       36.118028         62.0       40.0       45.5620   
4                       39.386040         68.0       44.0       54.2181   
5                       10.403038         15.0       11.0       12.1171   
6                       18.045481         44.0       17.0       21.2890   
7                       28.061896   

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

In [35]:
# df[['trip_uuid','source_center','destination_center']]
df.head()

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

                od_start_time  ... osrm_distance    factor  \
0  2018-09-20 03:21:32.418600  ...       11.9653  1.272727   
1  2018-09-20 03:21:32.418600  ...       21.7243  1.200000   
2  2018-09-20 03:21:32.418600  ...       32.5395  1.428571   
3  2018-09-20 03:21:32.418600  ...       45.5620  1.550000   
4  2018-09-20 03:21:32.418600  ...       54.2181  1.545455   

   segment_actual_time  segment_osrm_time segment_osrm_distance  \
0                 14.0               11.0               11.9653   
1                 10.0                9.0                9.7590   
2                 16.0                7.0               10.8152   
3                 21.0               12.0               13.0224   
4                  6.0                5.0                3.9153   

   segment_factor                                      segment_key  \
0        1.272727  trip-153741093647649320IND388121AAAIND388620AAB   
1        1.111111  trip-153741093647649320IND388121AAAIND388620AAB   
2        2.285714  trip-153741093647649320IND388121AAAIND388620AAB   
3        1.750000  trip-153741093647649320IND388121AAAIND388620AAB   
4        1.200000  trip-153741093647649320IND388121AAAIND388620AAB   

   segment_actual_time_sum  segment_osrm_distance_sum  segment_osrm_time_sum  
0                     14.0                    11.9653                   11.0  
1                     24.0                    21.7243                   20.0  
2                     40.0                    32.5395                   27.0  
3                     61.0                    45.5619                   39.0  
4                     67.0                    49.4772                   44.0  

[5 rows x 28 columns]

In [7]:
grouped = df.groupby(['source_name','destination_name']).aggregate({
    'actual_time':"last",
    'segment_actual_time':['sum']
})

print(grouped)

                                                                         actual_time  \
                                                                                last   
source_name                     destination_name                                       
AMD_Memnagar (Gujarat)          Ahmedabad_East_H_1 (Gujarat)                    37.0   
AMD_Rakhial (Gujarat)           Ahmedabad_East_H_1 (Gujarat)                    38.0   
Abohar_DC (Punjab)              Malout_DC (Punjab)                              19.0   
                                Muktsar_DPC (Punjab)                            98.0   
Achrol_BgwriDPP_D (Rajasthan)   Jaipur_Hub (Rajasthan)                          77.0   
...                                                                              ...   
Weir_BsstdDPP_D (Rajasthan)     Kherli_GodamDPP_D (Rajasthan)                   77.0   
YamunaNagar_DC (Haryana)        PaontSahib_Gurudwar_D (Himachal Pradesh)       104.0   
Yellandu_Sudimala_D (Telangana) 

In [9]:
df.head(10)

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   
5  training  2018-09-20 02:35:36.476840   
6  training  2018-09-20 02:35:36.476840   
7  training  2018-09-20 02:35:36.476840   
8  training  2018-09-20 02:35:36.476840   
9  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
5  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
6  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
7  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
8  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
9  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                    source_name  \
0  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA     Anand_VUNagar_DC (Gujarat)   
5  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
6  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
7  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
8  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
9  trip-153741093647649320  IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
5       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
6       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
7       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
8       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   
9       IND388320AAA     Anand_Vaghasi_IP (Gujarat)   

                od_start_time  ...            cutoff_timestamp  \
0  2018-09-20 03:21:32.418600  ...         2018-09-20 04:27:55   
1  2018-09-20 03:21:32.418600  ...         2018-09-20 04:17:55   
2  2018-09-20 03:21:32.418600  ...  2018-09-20 04:01:19.505586   
3  2018-09-20 03:21:32.418600  ...         2018-09-20 03:39:57   
4  2018-09-20 03:21:32.418600  ...         2018-09-20 03:33:55   
5  2018-09-20 04:47:45.236797  ...         2018-09-20 06:15:58   
6  2018-09-20 04:47:45.236797  ...         2018-09-20 05:47:29   
7  2018-09-20 04:47:45.236797  ...         2018-09-20 05:25:58   
8  2018-09-20 04:47:45.236797  ...         2018-09-20 05:15:56   
9  2018-09-20 04:47:45.236797  ...         2018-09-20 04:49:20   

   actual_distance_to_destination  actual_time  osrm_time osrm_distance  \
0                       10.435660         14.0       11.0       11.9653   
1                       18.936842         24.0       20.0       21.7243   
2                       27.637279         40.0       28.0       32.5395   
3                       36.118028         62.0       40.0       45.5620   
4                       39.386040         68.0       44.0       54.2181   
5                       10.403038         15.0       11.0       12.1171   
6                       18.045481         44.0       17.0       21.2890   
7                       28.061896   

In [16]:
df['segment_key'] = df['trip_uuid'] + df['source_center'] + df['destination_center']

# print(df['segment_key'])

segment_cols = ['segment_actual_time' , "segment_osrm_distance" ,"segment_osrm_time"]
print(segment_cols)


for col in segment_cols:
    df[col+'_sum'] = df.groupby('segment_key')[col].cumsum()
    
    

df[[col+'_sum' for col in segment_cols ]]

['segment_actual_time', 'segment_osrm_distance', 'segment_osrm_time']


segment_actual_time_sum  segment_osrm_distance_sum  \
0                          14.0                    11.9653   
1                          24.0                    21.7243   
2                          40.0                    32.5395   
3                          61.0                    45.5619   
4                          67.0                    49.4772   
...                         ...                        ...   
144862                     92.0                    65.3487   
144863                    118.0                    82.7212   
144864                    138.0                   103.4265   
144865                    155.0                   122.3150   
144866                    423.0                   131.1238   

        segment_osrm_time_sum  
0                        11.0  
1                        20.0  
2                        27.0  
3                        39.0  
4                        44.0  
...                       ...  
144862                   94.0  
144863                  115.0  
144864                  149.0  
144865                  176.0  
144866                  185.0  

[144867 rows x 3 columns]

In [53]:
# df.groupby(['trip_uuid','source_center','destination_center']).aggregate({
#     'actual_time':"last",
#     'segment_actual_time':"last"
# })


# df.groupby(['trip_uuid','source_center','destination_center'])['segment_actual_time'].cumsum()
df

trip_creation_time  year  month  day  hour  minute  second  \
0 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
1 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
2 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
3 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
4 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
5 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
6 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
7 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
8 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
9 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   

   microsecond  
0       476840  
1       476840  
2       476840  
3       476840  
4       476840  
5       476840  
6       476840  
7       476840  
8       476840  
9       476840

### Convert  trip creation time to year date time seconds and day like monday etc

# 2018-09-20 02:35:36.476840

In [54]:
from datetime import  datetime

In [55]:



# data = df['trip_creation_time']
# dft = pd.DataFrame(data)

# Convert the 'trip_creation_time' column to datetime
# dft['trip_creation_time'] = pd.to_datetime(data)

# Extract components
# dft['year'] = df['trip_creation_time'].dt.year
# dft['month'] = df['trip_creation_time'].dt.month
# dft['day'] = df['trip_creation_time'].dt.day
# dft['hour'] = df['trip_creation_time'].dt.hour
# dft['minute'] = df['trip_creation_time'].dt.minute
# dft['second'] = df['trip_creation_time'].dt.second

# Display the DataFrame
# print(dft)


In [62]:
df['year'] = df['trip_creation_time'].dt.year
df['day'] = df['trip_creation_time'].dt.day
df['hour'] = df['trip_creation_time'].dt.hour
df['minute'] = df['trip_creation_time'].dt.minute
df['second'] = df['trip_creation_time'].dt.second



df.head()


trip_creation_time  year  month  day  hour  minute  second  \
0 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
1 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
2 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
3 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   
4 2018-09-20 02:35:36.476840  2018      9   20     2      35      36   

   microsecond  
0       476840  
1       476840  
2       476840  
3       476840  
4       476840

In [63]:
df.groupby('trip_creation_time').count()

year  month  day  hour  minute  second  \
trip_creation_time                                                   
2018-09-20 02:35:36.476840    10     10   10    10      10      10   

                            microsecond  
trip_creation_time                       
2018-09-20 02:35:36.476840           10